RecoEgamma/EgammaTools/plugins/HGCalElectronIDValueMapProducer.cc

```diff
@@ -0,0 +1,276 @@
+  public:
+    explicit HGCalElectronIDValueMapProducer(const edm::ParameterSet&);
+    ~HGCalElectronIDValueMapProducer() override;
+
+    static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+  private:
+    void beginStream(edm::StreamID) override;
+    void produce(edm::Event&, const edm::EventSetup&) override;
+    void endStream() override;
+
+    // ----------member data ---------------------------
+    edm::EDGetTokenT<edm::View<reco::GsfElectron>> electronsToken_;
+    float radius_;
+
+    static const std::vector<std::string> valuesProduced_;
+    std::map<const std::string, std::vector<float>> maps_;
+
+    std::unique_ptr<HGCalEgammaIDHelper> eIDHelper_;
+};
+
+// All the ValueMap names to output are defined in the auto-generated python cfi
+// so that potential consumers can configure themselves in a simple manner
+// Would be cool to use compile-time validation, but need constexpr strings, e.g. std::string_view in C++17
+const std::vector<std::string> HGCalElectronIDValueMapProducer::valuesProduced_ = {
+    "ecOrigEt",
+    "ecOrigEnergy",
+    "ecEt",
+    "ecEnergy",
+    "ecEnergyEE",
+    "ecEnergyFH",
+    "ecEnergyBH",
+    "pcaEig1",
+    "pcaEig2",
+    "pcaEig3",
+    "pcaSig1",
+    "pcaSig2",
+    "pcaSig3",
+    "pcaAxisX",
+    "pcaAxisY",
+    "pcaAxisZ",
+    "pcaPositionX",
+    "pcaPositionY",
+    "pcaPositionZ",
+    "sigmaUU",
+    "sigmaVV",
+    "sigmaEE",
+    "sigmaPP",
+    "nLayers",
+    "firstLayer",
+    "lastLayer",
+    "e4oEtot",
+    "layerEfrac10",
+    "layerEfrac90",
+    "measuredDepth",
+    "expectedDepth",
+    "expectedSigma",
+    "depthCompatibility",
+    "caloIsoRing0",
+    "caloIsoRing1",
+    "caloIsoRing2",
+    "caloIsoRing3",
+    "caloIsoRing4",
+};
+
+HGCalElectronIDValueMapProducer::HGCalElectronIDValueMapProducer(const edm::ParameterSet& iConfig) :
+  electronsToken_(consumes<edm::View<reco::GsfElectron>>(iConfig.getParameter<edm::InputTag>("electrons"))),
+  radius_(iConfig.getParameter<double>("pcaRadius"))
+{
+  for(const auto& key : valuesProduced_) {
+    maps_[key] = {};
+    produces<edm::ValueMap<float>>(key);
+  }
+
+  eIDHelper_.reset(new HGCalEgammaIDHelper(iConfig, consumesCollector()));
+}
+
+
+HGCalElectronIDValueMapProducer::~HGCalElectronIDValueMapProducer()
+{
+}
+
+
+// ------------ method called to produce the data  ------------
+void
+HGCalElectronIDValueMapProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup)
+{
+  using namespace edm;
+
+  Handle<edm::View<reco::GsfElectron>> electronsH;
+  iEvent.getByToken(electronsToken_, electronsH);
+
+  // Clear previous map
+  for(auto&& kv : maps_) {
+    kv.second.clear();
+    kv.second.reserve(electronsH->size());
+  }
+
+  // Set up helper tool
+  eIDHelper_->eventInit(iEvent,iSetup);
+
+  for(const auto& electron : *electronsH) {
+    if(electron.isEB()) {
+      // Fill some dummy value
+      for(auto&& kv : maps_) {
+	kv.second.push_back(0.);
+      }
```

Comment:
is there a continue missing here? (or are barrel electrons meant to go through to make what seems to be hgcal specific outputs?

Reply:
The hgcal part is in the else statement

Reply:
Right - ah - i missed that the else continues for the rest of the loop.. Got it. Thanks